cpp/src/io/csv/csv_reader.cu

```diff
@@ -439,27 +441,60 @@ gdf_error read_csv(csv_read_arg *args)
 	if(args->names == nullptr) {
 		int h_num_cols = 0;
 
-		if (raw_csv->header_row >= 0) {
-			// Parse header row to assign column names; does not currently handle
-			// quotations to avoid added parsing complexity
-			for (size_t pos = 0, prev = 0; pos < raw_csv->header.size(); ++pos) {
-				if (raw_csv->header[pos] == raw_csv->opts.delimiter ||
-					raw_csv->header[pos] == raw_csv->opts.terminator) {
-					raw_csv->col_names.emplace_back(raw_csv->header.data() + prev, pos - prev);
-					h_num_cols++;
-					if (raw_csv->header[pos] == raw_csv->opts.terminator) {
-						break;
+		vector<char> data_row;
+		if (raw_csv->header.empty()) {
+			cu_recstart_t first_row_len{};
+			if (raw_csv->num_records > 1) {
+				CUDA_TRY(cudaMemcpy(&first_row_len, raw_csv->recStart + 1, sizeof(cu_recstart_t), cudaMemcpyDefault));
+			}
+			else {
+				first_row_len = raw_csv->num_bytes / sizeof(char);
+				std::cout << first_row_len << ' ';
+			}
+			data_row.resize(first_row_len);
+			CUDA_TRY(cudaMemcpy(data_row.data(), raw_csv->data, raw_csv->num_bytes, cudaMemcpyDefault));
+		}
+		// Parse header row to assign column names
+		const auto& first_row = !raw_csv->header.empty() ? raw_csv->header : data_row;
+		const string str_prefix = args->prefix != nullptr ? string(args->prefix) : "";
+		bool quotation	= false;
+		for (size_t pos = 0, prev = 0; pos < first_row.size(); ++pos) {
+			if(first_row[pos] == raw_csv->opts.quotechar &&
+			(!quotation || pos == 0 || first_row[pos - 1] != '\\')) {
+				quotation = !quotation;
+				continue;
+			}
+			if (!quotation &&
+				(first_row[pos] == raw_csv->opts.delimiter ||
+				first_row[pos] == raw_csv->opts.terminator)) {
+					if (raw_csv->header_row >= 0) {
```

Comment:
I don't really see how. These two cases now have almost all code in common, because I have to account for delim_whitespace in both cases. The case without header_row adds five lines in total to this code.

I can refactor some of the code into separate functions, to reduce the size of the loop.

Reply:
@mjsamoht Refactored the code a bit, hopefully it's clearer now.